server.go

```diff
@@ -0,0 +1,196 @@
+		pipelines         = s.registry.pipelines
+		pipelineResponses = []pipelineResponse{}
+	)
+
+	for name, pipeline := range pipelines {
+		phases := []phaseResponse{}
+		for phase, reconcilers := range pipeline.Phases() {
+			resources := []string{}
+			for _, r := range reconcilers {
+				resources = append(resources, r.Metadata().Name)
+			}
+
+			phases = append(phases, phaseResponse{
+				Name:      phase,
+				Resources: resources,
+			})
+		}
+
+		pipelineResponses = append(pipelineResponses, pipelineResponse{
+			Name:   name,
+			Phases: phases,
+		})
+	}
+
+	// TODO: handle pagination
+	json.NewEncoder(w).Encode(listPipelinesResponse{
+		Pipelines: pipelineResponses,
+	})
+}
+
+func (s *Server) getPipeline(w http.ResponseWriter, r *http.Request) {
+	pipelineName := chi.URLParam(r, "pipeline")
+	pipeline, ok := s.registry.pipelines[pipelineName]
+	if !ok {
+		http.Error(w, "pipeline not found", http.StatusNotFound)
+		return
+	}
+
+	phases := []phaseResponse{}
+	for phase, reconcilers := range pipeline.Phases() {
+		resources := []string{}
+		for _, r := range reconcilers {
+			resources = append(resources, r.Metadata().Name)
+		}
+
+		phases = append(phases, phaseResponse{
+			Name:      phase,
+			Resources: resources,
+		})
+	}
+
+	json.NewEncoder(w).Encode(pipelineResponse{
+		Name:   pipelineName,
+		Phases: phases,
+	})
+}
+
+func (s *Server) getPhase(w http.ResponseWriter, r *http.Request) {
+	var (
+		pipelineName = chi.URLParam(r, "pipeline")
+		phaseName    = chi.URLParam(r, "phase")
+	)
+
+	pipeline, ok := s.registry.pipelines[pipelineName]
+	if !ok {
+		http.Error(w, "pipeline not found", http.StatusNotFound)
+		return
+	}
+
+	phases := pipeline.Phases()
+	reconcilers, ok := phases[phaseName]
+	if !ok {
+		http.Error(w, "phase not found", http.StatusNotFound)
+		return
+	}
+
+	resources := []string{}
+	for _, r := range reconcilers {
+		resources = append(resources, r.Metadata().Name)
+	}
+
+	json.NewEncoder(w).Encode(phaseResponse{
+		Name:      phaseName,
+		Resources: resources,
+	})
+}
+
+func (s *Server) getResource(w http.ResponseWriter, r *http.Request) {
+	var (
+		pipelineName = chi.URLParam(r, "pipeline")
+		phaseName    = chi.URLParam(r, "phase")
+		resourceName = chi.URLParam(r, "resource")
+	)
+
+	pipeline, ok := s.registry.pipelines[pipelineName]
+	if !ok {
+		http.Error(w, "pipeline not found", http.StatusNotFound)
+		return
+	}
+
+	phases := pipeline.Phases()
+	reconcilers, ok := phases[phaseName]
+	if !ok {
+		http.Error(w, "phase not found", http.StatusNotFound)
+		return
+	}
+
+	var targetResource core.Reconciler
+	for _, reconciler := range reconcilers {
+		if reconciler.Metadata().Name == resourceName {
+			targetResource = reconciler
+			break
+		}
+	}
+
+	if targetResource == nil {
+		http.Error(w, "resource not found", http.StatusNotFound)
+		return
+	}
+
+	metadata := targetResource.Metadata()
+	json.NewEncoder(w).Encode(resourceResponse{
+		Name:   metadata.Name,
+		Labels: metadata.Labels,
+	})
```

Comment:
The second map should be keyed by metadata name already:
```suggestion
	reconciler, ok := reconcilers[resourceName]
	if !ok {
		http.Error(w, "resource not found", http.StatusNotFound)
		return
	}
	
	metadata := resource.Metadata()
	json.NewEncoder(w).Encode(resourceResponse{
		Name:   metadata.Name,
		Labels: metadata.Labels,
	})
```

Reply:
Also, we have the option here to add the payload to rhe resourceResponse:
```go
payload, err := resource.Get(r.Context())
if err != nil {
	http.Error(w, err.Error(), http.StatusInternalServerError)
	return
}

json.NewEncoder(w).Encode(resourceResponse{
	Name:    metadata.Name,
	Labels:  metadata.Labels,
	Payload: payload,
})
```